tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
+#define LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
+
+#include <complex>
+
+namespace KokkosKernelsSTD {
+
+namespace Impl {
+
+// manages parallel execution of independent action
+// called like action(i, j) for each matrix element A(i, j)
+template <typename ExecSpace, typename MatrixType>
+class ParallelMatrixVisitor {
+public:
+  KOKKOS_INLINE_FUNCTION ParallelMatrixVisitor(ExecSpace &&exec_in, MatrixType &A_in):
+    exec(exec_in), A(A_in), ext0(A.extent(0)), ext1(A.extent(1))
+  {}
+
+  template <typename ActionType>
+  void for_each_triangle_matrix_element(std::experimental::linalg::upper_triangle_t t, ActionType action) {
+    Kokkos::parallel_for(Kokkos::RangePolicy(exec, 0, ext1),
+      KOKKOS_LAMBDA(const auto j) {
+        using idx_type = std::remove_const_t<decltype(j)>;
+        for (idx_type i = 0; i <= j; ++i) {
+          action(i, j);
+        }
+      });
+    exec.fence();
+  }
+
+  template <typename ActionType>
+  void for_each_triangle_matrix_element(std::experimental::linalg::lower_triangle_t t, ActionType action) {
+    for_each_triangle_matrix_element(std::experimental::linalg::upper_triangle,
+        [action](const auto i, const auto j) {
+          action(j, i);
+      });
+  }
+
+private:
+
+  template <typename T>
+  KOKKOS_INLINE_FUNCTION
+  static bool is_int_mul_safe(T a, T b) {
+    static constexpr auto max_val = std::numeric_limits<decltype(a * b)>::max();
+    return a <= max_val / b;
```

Comment:
I'm sorry for trying to sneak that in... I removed `is_int_mul_safe()` on https://github.com/kokkos/stdBLAS/pull/167/commits/c209631bc02ba6038d105f69d6fbd2fe48004e13, because it's no longer needed here*, but your your corrected version is preserved for future use: thank you very much for super useful lesson on `constexpr`.

*I thought of fine grained parallelism with one matrix element per thread (instead of whole row/column) - which requires `extent0 * extent1` not to overflow, thus `is_int_mul_safe()` - but then realized it doesn't make any sense on CPU because of extra per-element cost on index splitting (`i = ij / extent0` and it's worse for triangles), so I dropped that route as premature optimization.